Converted risk level transaction to task (EXPOSUREAPP-2842)

chris-cwa opened the pull request:
Using CWAs new task controller for risk level calculation.

Change to user experience of the app is not intended and should keep this PR from being approved.

- [x]  unit tests for risk level task and RiskLevels class
- [x] check changes in https://github.com/corona-warn-app/cwa-app-android/pull/1418
- [x] wait for https://github.com/corona-warn-app/cwa-app-android/pull/1418 to be approved

chris-cwa committed:
converted risk level transaction to task

chris-cwa committed:
simplified

chris-cwa committed:
fixed config: timeout is now duration

chris-cwa committed:
less is more

ralfgehrer commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt:
```diff
@@ -0,0 +1,246 @@
+    companion object {
+
+        private var TAG = DefaultRiskLevels::class.simpleName
+
```
remove empty lines plz

ralfgehrer commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt:
```diff
@@ -0,0 +1,246 @@
+        attenuationParameters: ApplicationConfigurationOuterClass.AttenuationDuration,
+        exposureSummary: ExposureSummary
+    ): Double {
+
```
remove empty line plz

ralfgehrer commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt:
```diff
@@ -0,0 +1,246 @@
+        return round(riskScore.times(DECIMAL_MULTIPLIER)).div(DECIMAL_MULTIPLIER)
+    }
+
+    private fun Int.capped() =
```
personally, I prefer to have override functions first, then public, then private

ralfgehrer commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt:
```diff
@@ -0,0 +1,246 @@
+            }
+
+            // risk level calculation date update
+            val lastCalculatedRiskLevelDate = LocalData.lastTimeRiskLevelCalculation()
```
LocalData should be injected IMHO

ralfgehrer commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt:
```diff
@@ -0,0 +1,246 @@
+     * @param riskLevel
+     */
+    private fun updateRiskLevelScore(riskLevel: RiskLevel) {
+        val lastCalculatedScore = RiskLevelRepository.getLastCalculatedScore()
```
RiskLevelRepository should be injected

ralfgehrer commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt:
```diff
@@ -0,0 +1,246 @@
+        try {
+            Timber.tag(TAG).v("update the risk level with $riskLevel")
+            val lastCalculatedRiskLevelScoreForRollback =
+                RiskLevelRepository.getLastCalculatedScore()
```
RiskLevelRepository should be injected

chris-cwa replied:
will be done in another PR

ralfgehrer commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt:
```diff
@@ -0,0 +1,246 @@
+        val lastCalculatedScore = RiskLevelRepository.getLastCalculatedScore()
+        if (RiskLevel.riskLevelChangedBetweenLowAndHigh(lastCalculatedScore, riskLevel)) {
+            NotificationHelper.sendNotification(
+                CoronaWarnApplication.getAppContext().getString(R.string.notification_body),
```
rather use LazyString

chris-cwa replied:
will be done in another PR

chris-cwa committed:
arrangement + empty lines

chris-cwa committed:
Merge branch 'release/1.6.x' into feature/2909_refactor_risk_level_transaction

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt:
```diff
@@ -0,0 +1,243 @@
+     *
+     * @return the [ApplicationConfigurationOuterClass.ApplicationConfiguration] from the backend
+     */
+    private suspend fun getApplicationConfiguration(): ApplicationConfigurationOuterClass.ApplicationConfiguration =
```
Inject `AppConfigProvider`  and access the config from there please

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelTask.kt:
```diff
@@ -0,0 +1,135 @@
+
+class RiskLevelTask @Inject constructor(
+    private val riskLevels: RiskLevels,
+    private val context: Context
```
I think you need to annotate this with `@AppContext'

d4rken replied:
The `context` needs to be annotated, not the class.
Because there are multiple contexts through out an Android app we want to tell dagger which specific one we want.
It's just a custom "tag" we created with the name "AppContext", which we apply in `AndroidModule.kt`.

d4rken reviewed (changes requested):
Looking good

d4rken commented:
Didn't do an indepth review yet, just short lookthrough.

kolyaopahle committed:
Added additional checks to notification of exposure/test-result to check for finished submission

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

chris-cwa committed:
Merge branch 'release/1.6.x' into feature/2909_refactor_risk_level_transaction

chris-cwa committed:
Merge branch 'release/1.6.x' into feature/2909_refactor_risk_level_transaction

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelTask.kt:
```diff
@@ -0,0 +1,135 @@
+
+    override suspend fun run(arguments: Task.Arguments) = try {
+        Timber.d("Running with arguments=%s", arguments)
+        Result(
```
Don't nest a whole block like this please

chris-cwa replied:
OK.

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelTask.kt:
```diff
@@ -0,0 +1,135 @@
+                    }
+                }
+            } else {
+                RiskLevelRepository.setLastCalculatedRiskLevelAsCurrent()
```
Maybe early abort instead of the big if-else?

chris-cwa replied:
OK.

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelTask.kt:
```diff
@@ -0,0 +1,135 @@
+    }
+
+    companion object {
+        private val RISK_LEVEL_TRANSACTION_TIMEOUT = Duration.standardMinutes(8)
```
Improved readability if we inline this?

chris-cwa replied:
OK.

chris-cwa committed:
submit risk level task requests

chris-cwa committed:
- unreachable catch blocks

chris-cwa committed:
housekeeping

chris-cwa committed:
Merge branch 'fix/release1.6.x_housekeeping' into feature/2909_refactor_risk_level_transaction

chris-cwa committed:
satisfied ktlint + detekt

chris-cwa committed:
fixed flow: isRefreshing

chris-cwa committed:
Merge branch 'release/1.6.x' into feature/2909_refactor_risk_level_transaction

chris-cwa committed:
fixed flavour for testers

chris-cwa committed:
refactored benchmark

chris-cwa committed:
fixed package declaration

chris-cwa committed:
Merge branch 'release/1.6.x' into feature/2909_refactor_risk_level_transaction

chris-cwa committed:
unit tests

chris-cwa committed:
Merge branch 'feature/2909_refactor_risk_level_transaction' of https://github.com/corona-warn-app/cwa-app-android into feature/2909_refactor_risk_level_transaction

chris-cwa committed:
fixed injection

chris-cwa committed:
fixed tests

ralfgehrer commented on Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/RiskLevelAndKeyRetrievalBenchmark.kt:
```diff
@@ -29,7 +33,7 @@ class RiskLevelAndKeyRetrievalBenchmark(
     suspend fun start(
         callCount: Int,
-        callback: (resultInfo: String) -> Unit
+        callback: Callback
```
Can you please give "Callback" a more descriptive name

ralfgehrer commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt:
```diff
@@ -0,0 +1,234 @@
+
+    override fun calculationNotPossibleBecauseNoKeys() =
+        (TimeVariables.getLastTimeDiagnosisKeysFromServerFetch() == null).also {
+            if (it)
```
.also.if(true) ? Why not simplify

`if(TimeVariables.getLastTimeDiagnosisKeysFromServerFetch() == null) { Timber.... }`

Is the same, isn't it.

ralfgehrer commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt:
```diff
@@ -0,0 +1,234 @@
+        return round(riskScore.times(DECIMAL_MULTIPLIER)).div(DECIMAL_MULTIPLIER)
+    }
+
+    fun Int.capped() =
```
should be private (or internal with a @VisibleForTesting annotation)

ralfgehrer commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt:
```diff
@@ -0,0 +1,234 @@
+            this
+        }
+
+    fun withinDefinedLevelThreshold(riskScore: Double, min: Int, max: Int) =
```
should be private (or internal with a @VisibleForTesting annotation)

ralfgehrer commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt:
```diff
@@ -0,0 +1,234 @@
+     *
+     * @param riskLevel
+     */
+    fun updateRiskLevelScore(riskLevel: RiskLevel) {
```
should be private (or internal with a @VisibleForTesting annotation)

ralfgehrer commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt:
```diff
@@ -0,0 +1,234 @@
+        }
+    }
+
+    fun calculateRiskScore(
```
should be private (or internal with a @VisibleForTesting annotation)

ralfgehrer commented on Corona-Warn-App/src/test/java/de/rki/coronawarnapp/risk/RiskLevelsTest.kt:
```diff
@@ -1,16 +1,43 @@
+
+    @Test
+    fun `is within defined level threshold`() {
+        // positive
```
I'd prefer separate test cases, so it's obvious by the test case name which scenario failed, instead of having to look at the line number.

chris-cwa committed:
Merge branch 'release/1.6.x' into feature/2909_refactor_risk_level_transaction

harambasicluka committed:
Merge branch 'release/1.6.x' into feature/2909_refactor_risk_level_transaction

chris-cwa committed:
Merge branch 'release/1.6.x' into feature/2909_refactor_risk_level_transaction

chris-cwa committed:
comments from PR

chris-cwa committed:
Merge branch 'release/1.6.x' into feature/2909_refactor_risk_level_transaction

kolyaopahle committed:
Merge branch 'release/1.6.x' into fix/3182-3181-notifications-in-invalid-states

kolyaopahle commented:
i made some small changes to the risk transaction in https://github.com/corona-warn-app/cwa-app-android/pull/1418 please keep those in mind when updating this pr 👍

chris-cwa commented:
> i made some small changes to the risk transaction in #1418 please keep those in mind when updating this pr 👍

thank you!

chris-cwa committed:
Merge remote-tracking branch 'origin/release/1.6.x' into feature/2909_refactor_risk_level_transaction

# Conflicts:
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/TracingRepository.kt

chris-cwa committed:
Merge branch 'feature/2909_refactor_risk_level_transaction' of https://github.com/corona-warn-app/cwa-app-android into feature/2909_refactor_risk_level_transaction

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt:
```diff
@@ -0,0 +1,238 @@
+    override fun updateRepository(riskLevel: RiskLevel, time: Long) {
+        val rollbackItems = mutableListOf<RollbackItem>()
+        try {
+            Timber.tag(TAG).v("update the risk level with $riskLevel")
```
As a general thing: I'd really like it if log statements start with a capital letter, their are sentences after all.

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt:
```diff
@@ -0,0 +1,238 @@
+                LocalData.lastTimeRiskLevelCalculation(lastCalculatedRiskLevelDate)
+            }
+        } catch (error: Exception) {
+            Timber.tag(TAG).e(error)
```
Please always log errors with an error message, e.g.  "Updating the RiskLevelRepository failed."

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt:
```diff
@@ -0,0 +1,238 @@
+                Timber.tag(TAG).d("Initiate Rollback")
+                for (rollbackItem: RollbackItem in rollbackItems) rollbackItem.invoke()
+            } catch (rollbackException: Exception) {
+                Timber.tag(TAG).e(rollbackException)
```
Same here, e.g. "RiskLevelRepository rollback failed."

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt:
```diff
@@ -0,0 +1,238 @@
+
+    override fun calculationNotPossibleBecauseNoKeys() =
+        (TimeVariables.getLastTimeDiagnosisKeysFromServerFetch() == null).also {
+            if (it)
```
When formatting forces us to go multi-line, please add brackets to `if` statements.

chris-cwa committed:
Merge remote-tracking branch 'origin/fix/3182-3181-notifications-in-invalid-states' into feature/2909_refactor_risk_level_transaction

# Conflicts:
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/transaction/RiskLevelTransaction.kt

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt:
```diff
@@ -0,0 +1,238 @@
+
+    override suspend fun isIncreasedRisk(): Boolean {
+        val lastExposureSummary = getNewExposureSummary()
+        // retrieve application configuration
```
We only need comments for things that are not obvious.

We don't need to explain what `val appConfiguration = appConfigProvider.getAppConfig()` is doing, right?

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt:
```diff
@@ -0,0 +1,238 @@
+
+    override suspend fun calculationNotPossibleBecauseTracingIsOff() =
+        // this applies if tracing is not activated
+        !InternalExposureNotificationClient.asyncIsEnabled()
```
We can get this without static access now, try `ENFClient.isTracingEnabled.first()`.

If possible use the injected ENFClient with `ENFClient.internalClient` so we can remove more singleton access. `InternalExposureNotificationClient` will be gone in the future.

chris-cwa commented:
> i made some small changes to the risk transaction in #1418 please keep those in mind when updating this pr 👍

#1418 has been merged to this branch.

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/TracingRepository.kt:
```diff
@@ -56,15 +59,22 @@ class TracingRepository @Inject constructor(
-    val internalIsRefreshing = MutableStateFlow(false)
+    private val retrievingDiagnosisKeys = MutableStateFlow(false)
+    private val internalIsRefreshing =
+        retrievingDiagnosisKeys.combine(AppInjector.component.taskController.tasks) { retrievingDiagnosisKeys, tasks ->
```
Please don't use static access if the the whole class is already injected, you can get the taskController via constructor injection here.

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/TracingRepository.kt:
```diff
@@ -77,17 +87,15 @@ class TracingRepository @Inject constructor(
-                RiskLevelTransaction.start()
-            } catch (e: TransactionException) {
-                e.cause?.report(ExceptionCategory.EXPOSURENOTIFICATION)
+                AppInjector.component.taskController.submit(DefaultTaskRequest(RiskLevelTask::class))
```
Constructor injection.

d4rken reviewed (changes requested):
Got this Exception on launch:

```kotlin
E/AndroidRuntime: FATAL EXCEPTION: DefaultDispatcher-worker-3 @coroutine#80
    Process: de.rki.coronawarnapp.test, PID: 21703
    de.rki.coronawarnapp.task.MissingTaskFactoryException: No task factory mapped for class de.rki.coronawarnapp.task.common.DefaultTaskRequest (Kotlin reflection is not available)
        at de.rki.coronawarnapp.task.TaskController.submit(TaskController.kt:68)
        at de.rki.coronawarnapp.storage.TracingRepository$refreshRiskLevel$1.invokeSuspend(TracingRepository.kt:168)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
```

chris-cwa committed:
fixed injection

chris-cwa committed:
Merge remote-tracking branch 'origin/release/1.6.x' into feature/2909_refactor_risk_level_transaction

# Conflicts:
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevelCalculation.kt
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelCalculation.kt
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/LocalData.kt
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/TracingRepository.kt
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/transaction/RiskLevelTransaction.kt
#	Corona-Warn-App/src/test/java/de/rki/coronawarnapp/risk/RiskLevelsTest.kt
#	Corona-Warn-App/src/test/java/de/rki/coronawarnapp/transaction/RiskLevelTransactionTest.kt

chris-cwa committed:
fixed lint issues

chris-cwa committed:
fixed another merge conflict

chris-cwa committed:
- unused class

chris-cwa committed:
map task factory

chris-cwa committed:
Merge branch 'release/1.6.x' into feature/2909_refactor_risk_level_transaction

chris-cwa committed:
fulfilled comment style wishes

chris-cwa committed:
Merge branch 'feature/2909_refactor_risk_level_transaction' of https://github.com/corona-warn-app/cwa-app-android into feature/2909_refactor_risk_level_transaction

chris-cwa committed:
fixed tests suspended by changes from develop branch

chris-cwa committed:
Merge remote-tracking branch 'origin/release/1.6.x' into feature/2909_refactor_risk_level_transaction

# Conflicts:
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/transaction/RiskLevelTransaction.kt

chris-cwa committed:
Merge branch 'release/1.7.x' into feature/2909_refactor_risk_level_transaction

harambasicluka committed:
Merge branch 'release/1.7.x' into feature/2909_refactor_risk_level_transaction

kolyaopahle commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/ExposureStateUpdateWorker.kt:
```diff
@@ -7,39 +7,33 @@ import com.google.android.gms.common.api.ApiException
+                    Timber.v("exposure summary state updated: $it")
+                }
+
+            AppInjector.component.taskController.submit(DefaultTaskRequest(RiskLevelTask::class))
```
Couldn't we constructor inject this?

d4rken replied:
Can't yet, will make this possible soon.

d4rken replied:
Will be made possible by #1503

kolyaopahle commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt:
```diff
@@ -0,0 +1,249 @@
+        }
+
+        // get the high risk score class
+        val highRiskScoreClass =
+            riskScoreClassification.riskClassesList.find { it.label == "HIGH" }
+                ?: throw RiskLevelCalculationException(IllegalStateException("No high risk score class found"))
```
is there a more robust way to handle this?

chris-cwa replied:
What do you mean by 'robust'? An exception has to be thrown in order to abort the task. All exceptions are caught by the task controller.

kolyaopahle replied:
the exception is fine, i was just thinking about the whole process of resolving the highest risk class by comparing a list of instance properties to a string.

kolyaopahle replied:
kind of seems like this might fail if the classes are changed and we forget to update here, there is no real hard coupling that enforces compile time correctness

kolyaopahle replied:
i'll resolve this as we discussed a fix for it internally

kolyaopahle commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt:
```diff
@@ -0,0 +1,249 @@
+    }
+
+    override fun isActiveTracingTimeAboveThreshold(): Boolean {
+        val durationTracingIsActive = TimeVariables.getTimeActiveTracingDuration()
+        val durationTracingIsActiveThreshold =
+            TimeVariables.getMinActivatedTracingTime().toLong()
+
+        val activeTracingDurationInHours = durationTracingIsActive.millisecondsToHours()
```
i would maybe reorder these so that both variables are worked on in order before being compared. This is only personal preference though.

```suggestion
        val durationTracingIsActive = TimeVariables.getTimeActiveTracingDuration()
        val activeTracingDurationInHours = durationTracingIsActive.millisecondsToHours()
        
        val durationTracingIsActiveThreshold =
            TimeVariables.getMinActivatedTracingTime().toLong()
```

harambasicluka reviewed (commented):
Looks really nice and is easy to read.
I just would like to get clarification on the two comments :)

I tested it on an Samsung Galaxy S8 and on on OnePlus 8.
Look very good, but I just tested the manual mode and the riskscore is calculated when you navigate from one screen to another. In my opinion this shouldn't happen, I will retest it with the latest RC to see if that's caused by this PR or it's already in.
-> ok it's at least in 1.6.0/RC5 I'll create a bug ticket for this.

SamuraiKek commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt:
```diff
@@ -0,0 +1,249 @@
+            TimeVariables.getMaxStaleExposureRiskRange() && isActiveTracingTimeAboveThreshold()
+    }
+
+    override fun calculationNotPossibleBecauseNoKeys() =
```
I would rename this function to calculationNotPossibleBecauseOfNoKeys, to keep the naming consistent with the function above.

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevels.kt:
```diff
@@ -0,0 +1,31 @@
+     */
+    fun isActiveTracingTimeAboveThreshold(): Boolean
+
+    suspend fun isIncreasedRisk(): Boolean
```
What do you think about refactoring this such that the `DefaultRiskLevels` implementation is not responsible for both retrieving the data and working on it?

SamuraiKek commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt:
```diff
@@ -0,0 +1,249 @@
+        val lastExposureSummary = getNewExposureSummary()
+        val appConfiguration = appConfigProvider.getAppConfig()
+        Timber.tag(TAG).v("Retrieved configuration from backend")
+        // custom attenuation parameters to weight the attenuation
```
Weigh, not weight.

chris-cwa replied:
Should we then also rename the field `attenuationParameters.weights`?

SamuraiKek replied:
I don't think so. Weight is used for defining the "heaviness" of something, and the verb "to weigh" is used to find the weight of something. The attenuationParameters indeed have weights. Sorry I don't mean to sound like the grammar police :D

chris-cwa replied:
I didn't know that! Thank you so much!

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevels.kt:
```diff
@@ -0,0 +1,31 @@
+
+    suspend fun isIncreasedRisk(): Boolean
+
+    suspend fun calculationNotPossibleBecauseTracingIsOff(): Boolean
```
Moving things like early abort checks (e.g. `calculationNotPossibleBecauseTracingIsOff`) out of this interface? Into the task?

I think it's already A LOT better than before, but I have still trouble following the code path through the different steps. In my head I always try to imagine how the data is passed from one step to the next :thinking:.

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/TracingRepository.kt:
```diff
@@ -70,6 +78,10 @@ class TracingRepository @Inject constructor(
         }
     }
 
+    private fun List<TaskInfo>.isRiskLevelTaskRunning() = find {
```
I think there is a method called `List<T>.any { }` which returns true if an of the elements match the condition.

d4rken commented on Corona-Warn-App/src/test/java/de/rki/coronawarnapp/risk/RiskLevelsTest.kt:
```diff
@@ -1,16 +1,48 @@
 import org.junit.Test
 
-class RiskLevelCalculationTest {
+class RiskLevelsTest {
```
Please extend `BaseTest` which automatically set's up logging so that we can see log output if the classes produce any.

d4rken reviewed (commented):
Love it so far, thinking about whether it worth it to put some things into their own class. Maybe we can have a short to discuss some ideas, brainstorm a bit.

BMItter commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt:
```diff
@@ -0,0 +1,249 @@
+    }
+
+    companion object {
+        private var TAG = DefaultRiskLevels::class.simpleName
```
TAG should be final (val) I think..
I also like private val TAG = DefaultRiskLevels::class.java.simpleName very much, because you get a String instead of String?, but please, don't change it if you don't want (just a personal note)

BMItter commented:
It's really a pleasure not to see the old Transactions anymore! Great Work! :heart:

chris-cwa commented:
> It's really a pleasure not to see the old Transactions anymore! Great Work! ❤️

Thanks! 😃

chris-cwa committed:
Update Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt

https://github.com/corona-warn-app/cwa-app-android/pull/1412#discussion_r514325776

Co-authored-by: Kolya Opahle <k.opahle@sap.com>

chris-cwa commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt:
```diff
@@ -0,0 +1,249 @@
+            Timber.tag(TAG)
+                .v("$riskScore is above the defined min value ${highRiskScoreClass.min}")
+            return true
+        } else if (riskScore > highRiskScoreClass.max) {
```
it was that day before, I tried to no change logic. I guess invalid values should not lead to a result

chris-cwa commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt:
```diff
@@ -0,0 +1,249 @@
+                riskLevel
+            ) && !LocalData.submissionWasSuccessful()
+        ) {
+            NotificationHelper.sendNotification(
```
This also was not changed to the way it behaved before. I you want I can change it.

harambasicluka replied:
It was just an idea, I wouldn't insist on that :)

chris-cwa committed:
changes from PR

chris-cwa committed:
fixed a test

chris-cwa committed:
show logs, ktlint issues

chris-cwa committed:
made log tag non-nullable

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelTask.kt:
```diff
@@ -0,0 +1,166 @@
+            with(riskLevels) {
+                return Result(
+                    when {
+                        !enfClient.isTracingEnabled.first().also {
```
Can we turn this into an early abort condition like the above pattern used with `isNetworkEnabled`?

chris-cwa replied:
good idea

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelTask.kt:
```diff
@@ -0,0 +1,166 @@
+                        calculationNotPossibleBecauseOfOutdatedResults().also {
+                            checkCancel()
+                        } -> if (backgroundJobsEnabled)
+                            UNKNOWN_RISK_OUTDATED_RESULTS
```
No multiline `if` without braces please.

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelTask.kt:
```diff
@@ -0,0 +1,166 @@
+                RiskLevelRepository.setLastCalculatedRiskLevelAsCurrent()
+                return Result(UNDETERMINED)
+            }
+            with(riskLevels) {
```
What about removing one level of indentation here?

```kotlin

return when {
....
}
.let { Result(it) }
.also { ... }


```

chris-cwa replied:
I believe it reduces the good overview that you quickly get in the current state by just looking at the code.

harambasicluka replied:
Agree with @chris-cwa currently it's just nice to read.

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelTask.kt:
```diff
@@ -0,0 +1,166 @@
+                        else -> LOW_LEVEL_RISK
+                    }.also {
+                        checkCancel()
+                        updateRepository(it, System.currentTimeMillis())
```
Please inject and use `TimeStamper` to improve testability.

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelTask.kt:
```diff
@@ -0,0 +1,166 @@
+    }
+
+    private val backgroundJobsEnabled: Boolean
+        get() = ConnectivityHelper.autoModeEnabled(CoronaWarnApplication.getAppContext()).also {
```
We have a non static variant available for that, see `BackgroundModeStatus`, Could we use that here?

chris-cwa committed:
improvements from reviewers

chris-cwa committed:
reduced static access

chris-cwa committed:
Merge remote-tracking branch 'origin/release/1.7.x' into feature/2909_refactor_risk_level_transaction

# Conflicts:
#	Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/ExposureStateUpdateWorker.kt

chris-cwa committed:
Saint Pipeline, please give me the green light

chris-cwa committed:
Merge remote-tracking branch 'origin/release/1.7.x' into feature/2909_refactor_risk_level_transaction

# Conflicts:
#	Corona-Warn-App/src/test/java/de/rki/coronawarnapp/transaction/RiskLevelTransactionTest.kt

chris-cwa committed:
Merge branch 'release/1.7.x' into feature/2909_refactor_risk_level_transaction

AlexanderAlferov commented:
Tested. Looks good.
But I confirm Luka's comment regarding manual mode and risk auto-update when navigating from one screen to another.

chris-cwa committed:
toString()

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/ExposureStateUpdateWorker.kt:
```diff
@@ -9,42 +9,36 @@ import com.squareup.inject.assisted.Assisted
+                    Timber.v("exposure summary state updated: $it")
+                }
+
+            AppInjector.component.taskController.submit(DefaultTaskRequest(RiskLevelTask::class))
```
Inject the taskController

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/TracingRepository.kt:
```diff
@@ -162,16 +172,9 @@ class TracingRepository @Inject constructor(
-            } catch (e: Exception) {
-                e.report(ExceptionCategory.INTERNAL)
-            }
+            AppInjector.component.taskController.submit(DefaultTaskRequest(RiskLevelTask::class))
```
Injecet the taskController.

d4rken reviewed (dismissed):
Two cases where we can inject instead of static access, but otherwise looks good.

Successfully tested on a Pixel 3 @ Android 11

chris-cwa committed:
ktlint + di

chris-cwa committed:
fixed test

chris-cwa committed:
Merge branch 'release/1.7.x' into feature/2909_refactor_risk_level_transaction

d4rken reviewed (approved):
lgtm

ralfgehrer commented:
Successfully tested on Pixel (Android 10) and Pixel 5 (Android 11)
